document-store/src/main/java/org/hypertrace/core/documentstore/mongo/MongoCollection.java

```diff
@@ -123,13 +124,36 @@ public boolean upsert(Key key, Document document) throws IOException {
         LOGGER.debug("Write result: " + writeResult.toString());
       }
 
-      return writeResult.getModifiedCount() > 0;
+      return (writeResult.getUpsertedId() != null || writeResult.getModifiedCount() > 0);
     } catch (IOException e) {
       LOGGER.error("Exception upserting document. key: {} content:{}", key, document, e);
       throw e;
     }
   }
 
+  /**
+   * Same as existing upsert method, however, extends the support with condition and optional
+   * parameter for explicitly controlling insert and update.
+   * */
+  @Override
+  public boolean upsert(Key key, Document document, Filter condition, @Nullable Boolean isUpsert)
+      throws IOException {
+    try {
+      Map<String, Object> map = parseQuery(condition);
+      map.put(ID_KEY, key.toString());
+      BasicDBObject conditionObject = new BasicDBObject(map);
+      UpdateOptions options = new UpdateOptions().upsert(isUpsert != null ? isUpsert : true);
+      UpdateResult writeResult =
+          collection.updateOne(conditionObject, this.prepareUpsertWithCondition(key, document), options);
+      if (LOGGER.isDebugEnabled()) {
+        LOGGER.debug("Write result: " + writeResult.toString());
+      }
+      return (writeResult.getUpsertedId() == null || writeResult.getModifiedCount() > 0);
```

Comment:
What is success? Is an error or void return sufficient? Given the assumed intention, this looks like a bug.